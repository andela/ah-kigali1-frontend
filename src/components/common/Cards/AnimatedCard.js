import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import NoImg from "../../../assets/img/no-thumbnail.png";
import NoAvatar from "../../../assets/img/user.png";
import LikeIcon from "../../../assets/img/like-icon.svg";
import DisLikeIcon from "../../../assets/img/dislike-icon.svg";
import CommentIcon from "../../../assets/img/comment-icon.svg";
import BookmarkIcon from "../../../assets/img/bookmark-icons.svg";

export const AnimatedCard = ({
  title,
  likes,
  comments,
  description,
  author,
  readTime,
  createdAt,
  slug
}) => (
  <div className="blog-card">
    <div className="meta" data-test="thumbnail">
      <div
        className="article-thumbnail"
        style={{
          backgroundImage: `url(${NoImg})`
        }}
      />
    </div>
    <div className="card-content">
      <div className="info" data-test="author-info">
        <img
          src={author.image || NoAvatar}
          alt="user"
          className="user-profile-img"
        />
        <div className="more-info">
          <h3 className="username">{author.firstName || author.username}</h3>
          <span>
            {createdAt} {readTime} min read
          </span>
        </div>
      </div>
      <div className="card-title" data-test="article-details">
        <h2>{title}</h2>
      </div>
      <p className="card-text">
        {description}......
        <Link to={`/articles/${slug}`} className="read-more">
          Read more
        </Link>
      </p>
      <div className="user-actions">
        <div className="push-left">
          <div className="icon-action" data-test="action-icon">
            <img src={LikeIcon} alt="like" className="icon-like" />
            <span>{likes}</span>
          </div>
          <div className="icon-action" data-test="action-icon">
            <img src={DisLikeIcon} alt="dislike" className="icon-dislike" />
            <span>898</span>
          </div>
          <div className="icon-action" data-test="action-icon">
            <img src={CommentIcon} alt="like" className="icon-comment" />
            <span>{comments}</span>
          </div>
        </div>
        <div className="push-right">
          <div className="icon-action" data-test="action-icon">
            <img src={BookmarkIcon} alt="bookmark" className="icon-bookmark" />
          </div>
        </div>
      </div>
    </div>
  </div>
);

AnimatedCard.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  likes: PropTypes.number.isRequired,
  comments: PropTypes.number.isRequired,
  author: PropTypes.shape({
    following: PropTypes.bool,
    username: PropTypes.string,
    image: PropTypes.string
  }).isRequired,
  readTime: PropTypes.number.isRequired,
  createdAt: PropTypes.string.isRequired,
  slug: PropTypes.string
};
AnimatedCard.defaultProps = {
  slug: ""
};
export default AnimatedCard;
