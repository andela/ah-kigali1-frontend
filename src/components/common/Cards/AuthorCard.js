import React from "react";
import PropTypes from "prop-types";
import BasicButton from "../Buttons/BasicButton";
import defaultAvatar from "../../../assets/img/user.png";

const AuthorCard = props => {
  const { following, username, image } = props;
  return (
    <div className="author">
      <div className="inline-author">
        <div className="user-profile-img">
          <img src={image || defaultAvatar} alt="author" />
        </div>
        <p className="username">@{username}</p>
      </div>
      <span />
      <p className="author-bio">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </p>
      <BasicButton
        className={`following-btn ${following ? "btn-fill" : ""}`}
        title="Follow"
        onClick={() => null}
      />
    </div>
  );
};
AuthorCard.propTypes = {
  following: PropTypes.bool,
  image: PropTypes.string,
  username: PropTypes.string.isRequired
};
AuthorCard.defaultProps = {
  following: false,
  image: null
};
export default AuthorCard;
