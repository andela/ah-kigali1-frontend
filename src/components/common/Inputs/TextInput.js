import React from "react";
import PropTypes from "prop-types";

const TextInput = ({
  name,
  type,
  id,
  placeholder,
  value,
  onChange,
  ...props
}) => (
  <input
    {...props}
    name={name}
    type={type}
    id={id}
    placeholder={placeholder}
    value={value}
    onChange={onChange}
  />
);

TextInput.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  id: PropTypes.string,
  placeholder: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: "",
  onKeyDown: () => ""
};

export default TextInput;
