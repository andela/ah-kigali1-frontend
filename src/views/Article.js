import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import addArticle from "../redux/actions/index";
// import "../index.css";

const mapDispatchToProps = dispatch => ({
  addArticle: () => dispatch(addArticle())
});

const FetchedArticles = props => (
  <div>
    <h1>Fetch some articles from dummy app</h1>

    <button
      // eslint-disable-next-line react/destructuring-assignment
      onClick={() => props.addArticle()}
      className="btn"
      type="submit"
    >
      Fetch more articles
    </button>
  </div>
);

FetchedArticles.propTypes = {
  // eslint-disable-next-line react/require-default-props
  addArticle: PropTypes.func
};

const Articles = connect(
  null,
  mapDispatchToProps
)(FetchedArticles);

export default Articles;
