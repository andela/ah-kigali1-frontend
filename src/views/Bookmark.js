import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import BookmarkComponent from "../components/common/Bookmarks";
import bookmarkIcon from "../assets/img/bookmark-icons.svg";
import bookmarkedIcon from "../assets/img/bookmarked.svg";
import { handleBookmark } from "../redux/actions/bookmarkActions";

export class Bookmark extends Component {
  static chooseBookmarkIcon(condition) {
    return condition ? bookmarkedIcon : bookmarkIcon;
  }

  render() {
    const {
      bookmark,
      isBookmarked,
      slug,
      isSubmitting,
      bookmarkedArticles,
      location,
      history
    } = this.props;
    if (!Object.prototype.hasOwnProperty.call(bookmarkedArticles, slug)) {
      Object.assign(bookmarkedArticles, { [slug]: isBookmarked });
    }
    const theBookmarkIcon = Bookmark.chooseBookmarkIcon(
      bookmarkedArticles[slug]
    );
    return (
      <BookmarkComponent
        onClick={() =>
          bookmark(slug, bookmarkedArticles[slug], { location, history })
        }
        icon={theBookmarkIcon}
        disabled={isSubmitting}
      />
    );
  }
}

export const mapStateToProps = state => ({
  ...state.bookmark
});

export const mapDispatchToProps = dispatch => ({
  bookmark: (slug, isBookmarked, { history, location }) =>
    dispatch(handleBookmark(slug, isBookmarked, { history, location }))
});

Bookmark.propTypes = {
  bookmark: PropTypes.func.isRequired,
  isBookmarked: PropTypes.bool.isRequired,
  bookmarkedArticles: PropTypes.shape({}).isRequired,
  slug: PropTypes.string.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  location: PropTypes.shape([]).isRequired,
  history: PropTypes.shape({}).isRequired
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Bookmark)
);
