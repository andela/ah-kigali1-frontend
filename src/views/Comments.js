/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable no-shadow */
import React, { Component } from "react";
import Textarea from "react-textarea-autosize";
import PropTypes from "prop-types";
import moment from "moment";
import authorImage from "../assets/img/user.jpg";
import Button from "../components/common/Buttons/BasicButton";
import DeleteDialogue from "../components/common/Dialog/DeleteDialogue";
import heartIcon from "../assets/img/heart.svg";

export class Comments extends Component {
  state = {
    editMode: false,
    index: null
  };

  onEnterPress = (e, id) => {
    const { editComment } = this.props;
    if (e.keyCode === 13 && e.shiftKey === false) {
      e.preventDefault();
      editComment(id);
      this.setState({ editMode: false });
    }
  };

  isEditing = (body, index) => {
    const { onSetBodyEdit } = this.props;
    this.setState({ editMode: true, index });
    onSetBodyEdit({ value: body });
  };

  handleComment = id => {
    const { editComment } = this.props;
    editComment(id);
    this.setState({ editMode: false });
  };

  render() {
    const { editMode, index: commentIndex } = this.state;
    const {
      comments,
      updatedBody,
      deleteComment,
      handleCommentsInputEdit
    } = this.props;

    return Object.values(comments).length
      ? Object.values(comments).map((comment, index) => {
          const { body, like, id, createdAt, author } = comment;
          const { image, lastName, firstName, username } = author;
          return (
            <div className="article-comments--existing__desktop" key={id}>
              <div className="avatar-wrapper comment-avatar-wrapper">
                <img
                  src={image || authorImage}
                  alt="Avatar"
                  className="avatar"
                />
                <span className="comment-author_name ">
                  {username && firstName && lastName
                    ? `${firstName} ${lastName.charAt(0) ||
                        username.charAt(0)}.`
                    : username}
                </span>
              </div>
              <div className="article-comments--existing-text">
                {editMode && Number(commentIndex) === index ? (
                  <div>
                    <Textarea
                      id="edit-comment-textarea"
                      type="textarea"
                      onChange={handleCommentsInputEdit}
                      onKeyDown={e => this.onEnterPress(e, id)}
                      name="bodyEdit"
                      value={updatedBody}
                    />
                    <Button
                      id="update-comment"
                      className="btn delete_article"
                      onClick={() => this.handleComment(id)}
                      title="Update"
                    />
                    <Button
                      id="close-update-comment"
                      className="btn delete_article"
                      onClick={() => this.setState({ editMode: false })}
                      title="Cancel"
                    />
                  </div>
                ) : (
                  <div>
                    <div className="time-ago">
                      Posted {moment(createdAt).fromNow()}
                    </div>
                    <div className="comment-body" key={id}>
                      {body}
                    </div>
                    <div className="comment-edit-delete">
                      <DeleteDialogue
                        id="delete-comment"
                        deleteComment={() => deleteComment(id)}
                      />
                      <div
                        className="edit-comment"
                        onClick={() => this.isEditing(body, index)}
                      >
                        <i className="fa fa-edit" />
                      </div>
                    </div>
                    <div className="article-comments--actions">
                      <span className="comment-like">
                        <div className="icons">
                          <img src={heartIcon} alt="likes" className="likes" />
                          <div>{like}</div>
                        </div>
                      </span>
                    </div>
                  </div>
                )}
              </div>
            </div>
          );
        })
      : "";
  }
}

Comments.propTypes = {
  comments: PropTypes.shape([]).isRequired,
  updatedBody: PropTypes.string.isRequired,
  onSetBodyEdit: PropTypes.func.isRequired,
  deleteComment: PropTypes.func.isRequired,
  editComment: PropTypes.func.isRequired,
  handleCommentsInputEdit: PropTypes.func.isRequired
};

export default Comments;
