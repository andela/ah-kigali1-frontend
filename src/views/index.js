import React from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import Login from "./Login";
import ResetPassword from "./ResetPassword";
import Home from "./Home";
import UpdatePassword from "./UpdatePassword";
import SocialAuth from "./SocialAuth";
import NewArticle from "./NewArticle";
import EditArticle from "./EditArticle";
import NotFound from "./NotFound";
import ReadArticle from "./ReadArticle";

const Routers = () => (
  <Router>
    <Switch>
      <Route exact path="/" component={Home} />
      <Route exact path="/sign_in" component={Login} />
      <Route exact path="/social_auth" component={SocialAuth} />
      <Route exact path="/reset_password" component={ResetPassword} />
      <Route exact path="/update_password" component={UpdatePassword} />
      <Route exact path="/articles/:slug/edit" component={EditArticle} />
      <Route exact path="/articles/new_article" component={NewArticle} />
      <Route exact path="/not_found" component={NotFound} />
      <Route exact path="/articles/:slug" component={ReadArticle} />
    </Switch>
  </Router>
);
export default Routers;
