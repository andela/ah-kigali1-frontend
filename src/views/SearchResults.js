/* eslint-disable react/forbid-prop-types */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import queryString from "query-string";
import _ from "lodash";
import BasicButton from "../components/common/Buttons/BasicButton";
import Input from "../components/common/Inputs/TextInput";
import {
  handleInputChange,
  fetchResults,
  clearSearchResults
} from "../redux/actions/searchActions";
import {
  toReadableDate,
  filterByTag,
  getTags,
  isEmpty,
  isBottom
} from "../utils/helperFunctions";
import Card from "../components/common/Cards/AnimatedCard";
import AuthorCard from "../components/common/Cards/AuthorCard";
import Loading from "../components/Animations/LoadingDots";

export class SearchResults extends Component {
  state = {
    articles: [],
    authors: [],
    pageNumber: 1,
    activeTag: null,
    tagsList: []
  };

  componentWillMount() {
    const { location, fetchResults: getAllArticles, articles } = this.props;
    const { keyword } = queryString.parse(location.search);
    this.handleOnChange(keyword);
    if (isEmpty(articles)) getAllArticles(keyword, 1);
  }

  componentDidMount() {
    document.addEventListener("scroll", () => this.handleScroll(), true);
  }

  componentWillReceiveProps(nextProps) {
    const { articles, authors } = nextProps;
    this.setState({
      articles: _.values(articles),
      authors: _.values(authors),
      tagsList: getTags(articles)
    });
  }

  componentWillUnmount() {
    const { clearSearchResults: clear } = this.props;
    clear();
  }

  handleOnChange = value => {
    const { handleInputChange: setSearchQuery } = this.props;
    setSearchQuery(value);
  };

  handleEnterPress = e => {
    const { isLoading, searchQuery } = this.props;
    if (isLoading || isEmpty(searchQuery)) {
      return;
    }
    this.setState({
      pageNumber: 1
    });
    if (e.keyCode === 13 && e.shiftKey === false) {
      this.searchArticle(searchQuery);
    }
  };

  handleScroll = () => {
    console.log("***************", "scrolling");
    const { searchQuery, isLoading } = this.props;
    if (isBottom() && !isLoading) {
      this.setState(state => ({
        pageNumber: state.pageNumber + 1
      }));
      setTimeout(this.searchArticle(searchQuery), 3000);
    }
  };

  searchArticle = searchQuery => {
    const { fetchResults: getResults } = this.props;
    const { pageNumber } = this.state;
    getResults(searchQuery, pageNumber);
  };

  handleTagFilter = (tag, index) => {
    const { articles } = this.props;
    const { activeTag } = this.state;
    if (index === activeTag) {
      return this.setState({
        activeTag: null,
        articles: _.values(articles)
      });
    }
    this.setState({
      articles: filterByTag(_.values(articles), tag),
      activeTag: index
    });
  };

  render() {
    const { searchQuery, isLoading, errors } = this.props;
    const { articles, authors, activeTag, tagsList } = this.state;
    return (
      <div className="search-results">
        <div className="left-side">
          <div className="search-query">
            <Input
              type="search"
              name="search"
              placeholder=""
              className="search-input"
              onChange={e => this.handleOnChange(e.target.value)}
              value={searchQuery}
              onKeyDown={e => this.handleEnterPress(e)}
            />
          </div>
          <div className="results" id="results-container">
            <section id="articles" className="active">
              {articles.map(article => (
                <div className="col-md-12" key={article.id}>
                  <Card
                    title={article.title}
                    description={article.description}
                    comments={article.comments.length}
                    likes={article.likes.length}
                    author={article.author}
                    readTime={article.readTime}
                    createdAt={toReadableDate(article.createdAt)}
                  />
                </div>
              ))}
              <div className="loading-anim">
                {isLoading ? (
                  <Loading />
                ) : (
                  !isEmpty(errors.message) &&
                  !activeTag && <small>{errors.message}</small>
                )}
              </div>
            </section>
          </div>
        </div>

        <div className="right-side hide-sm">
          {tagsList.length !== 0 && <h2>#Tags</h2>}
          <div className="tags-section">
            {tagsList.map((title, index) => (
              <BasicButton
                className={`tags ${activeTag === index ? "active" : ""}`}
                title={title}
                key={`${index + 1}`}
                onClick={() => this.handleTagFilter(title, index)}
                data-test="single-tag"
              />
            ))}
          </div>
          {authors.length !== 0 && <h2>Authors</h2>}
          <div className="authors-section">
            {authors.map(author => (
              <AuthorCard {...author} key={author.id} />
            ))}
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => {
  const { search } = state;
  return {
    ...search
  };
};
SearchResults.propTypes = {
  handleInputChange: PropTypes.func.isRequired,
  searchQuery: PropTypes.string.isRequired,
  fetchResults: PropTypes.func.isRequired,
  articles: PropTypes.object.isRequired,
  authors: PropTypes.object.isRequired,
  location: PropTypes.shape({
    search: PropTypes.string
  }).isRequired,
  errors: PropTypes.shape({
    message: PropTypes.string
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  clearSearchResults: PropTypes.func.isRequired
};

export default connect(
  mapStateToProps,
  { handleInputChange, fetchResults, clearSearchResults }
)(SearchResults);
