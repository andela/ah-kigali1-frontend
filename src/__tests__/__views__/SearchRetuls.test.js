import React from "react";
import toJson from "enzyme-to-json";
import _ from "lodash";
import { shallow, mount } from "enzyme";
import { SearchResults } from "../../views/SearchResults";
import AnimatedCard from "../../components/common/Cards/AnimatedCard";
import TextInput from "../../components/common/Inputs/TextInput";
import { articles, authors } from "../testData";
import AuthorCard from "../../components/common/Cards/AuthorCard";

const [handleInputChange, fetchResults, clearSearchResults, push] = new Array(
  4
).fill(jest.fn());
const props = {
  handleInputChange,
  fetchResults,
  clearSearchResults,
  searchQuery: "hello_world",
  articles: {},
  authors: {},
  history: {
    location: {
      search: "?keyword=helloWorld"
    },
    push
  },
  isLoading: true,
  errors: {}
};
const wrapper = shallow(<SearchResults {...props} />);
const findElements = element => wrapper.find(element);
describe("Search Results Component", () => {
  afterEach(() => {
    handleInputChange.mockClear();
    fetchResults.mockClear();
  });
  describe("rendered component", () => {
    test("should match the snapshot", () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    test("should render right number of TextInput", () => {
      expect(findElements(TextInput).length).toBe(1);
    });
    test("should not render article card if isLoading true", () => {
      expect(findElements(AnimatedCard).length).toBe(0);
    });
    test("should render the right number of article card", () => {
      wrapper.setProps({
        articles: _.keyBy(articles, article => article.id),
        isLoading: false,
        authors: _.keyBy(authors, author => author.id)
      });
      expect(findElements(AnimatedCard).length).toBe(3);
      expect(findElements(AuthorCard).length).toBe(1);
    });
  });
  describe("component methods & functions", () => {
    let instance;
    beforeEach(() => {
      instance = wrapper.instance();
      jest.spyOn(instance, "handleOnChange");
      jest.spyOn(instance, "handleEnterPress");
    });
    afterEach(() => {
      instance.handleOnChange.mockClear();
      instance.handleEnterPress.mockClear();
    });

    test("should respond on change input text", () => {
      const value = "hello_world";
      findElements(TextInput)
        .at(0)
        .simulate("change", {
          target: {
            value
          }
        });
      expect(instance.handleOnChange).toHaveBeenCalledWith(value);
      expect(handleInputChange).toHaveBeenCalledWith(value);
    });
    test("should response on ENTER key press", () => {
      const searchQuery = "hello_world";
      findElements(TextInput)
        .at(0)
        .simulate("keydown", { keyCode: 13, shiftKey: false });
      expect(instance.handleEnterPress).toHaveBeenCalledWith({
        keyCode: 13,
        shiftKey: false
      });
      expect(fetchResults).toHaveBeenCalledWith(searchQuery, 1, props.history);
      expect(instance.searchArticle).toHaveBeenCalledWith(props.searchQuery);
    });
    test("should filter article by tags", () => {
      wrapper.setProps({
        articles: _.keyBy(articles, "id")
      });
      wrapper
        .find(`[data-test="single-tag"]`)
        .at(0)
        .simulate("click");
      expect(wrapper.state().activeTag).toEqual(0);
      expect(instance.handleTagFilter).toHaveBeenCalledWith(
        wrapper.state().tagsList[0],
        0
      );
    });
    describe("Component life cycle metho", () => {
      const component = mount(<SearchResults {...props} />);
      test("should fetch article and set search query", () => {
        expect(fetchResults).toHaveBeenCalled();
        expect(handleInputChange).toBeCalled();
      });
      test("should clear test result on component unmount", () => {
        component.unmount();
        expect(clearSearchResults).toBeCalled();
      });
    });
  });
});
