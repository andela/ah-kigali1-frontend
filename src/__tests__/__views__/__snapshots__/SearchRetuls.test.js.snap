// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Results Component rendered component should match the snapshot 1`] = `
<div
  className="search-results"
>
  <div
    className="left-side"
  >
    <div
      className="search-query"
    >
      <TextInput
        className="search-input"
        id=""
        name="search"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        type="search"
        value="helloWorld"
      />
    </div>
    <div
      className="results"
      id="results-container"
    >
      <section
        className="active"
        id="articles"
      >
        <div
          className="loading-anim"
        >
          <BottomPopUp />
        </div>
      </section>
    </div>
  </div>
  <div
    className="right-side hide-sm"
  />
</div>
`;
